### PR TITLE
WIP: MPI-ify the dynamics

### LINKED ISSUE #120: MPI parallelization of Dynamics

### PR DESCRIPTION
**Pull request by TomMelt**
This PR has been superseded by #744

Fixes #120

### Change Description

To parallelize the dynamics we first need the halo region information for each process.

This information is located in the metadata file produced by the domain `decomp` tool.

This PR updates `ModelMetadata::getPartitionMetadata` with all the necessary information on halo regions.

### Test Description

This PR adds two new tests to check the functionality of the `getPartitionMetadata` method and the halo exchange.
- [x] `ModelMetadata_test.cpp` - test `getPartitionMetadata` can correctly load metadata from a partition metadata file into the appropriate data structure e.g., `neighbourExtents`, `neighbourHaloStarts` etc.
- [x] `HaloExchange_test.cpp` - check that we can do basic halo exchange by making use of the new slicing utilities e.g., `ModelArraySlice` #744 
- [ ] add halo exchange logic to nextsim

### Pre-Request Checklist

- [x] Add methods to read halo exchange info into metadata class
- [ ] Add halo exchange info to DGModelArray class
- [ ] Add exchange logic to DGModelArray.hpp
- [ ] Add option of in-/out-flows in the halo? or should this be separate?

### CODE REVIEW

File: core/test/ModelMetadata_test.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+        REQUIRE(metadata.neighbourRanks[ModelMetadata::LEFT] == std::vector<int> { 0, 1 });
+        REQUIRE(metadata.neighbourExtents[ModelMetadata::LEFT] == std::vector<int> { 4, 5 });
+        REQUIRE(metadata.neighbourHaloStarts[ModelMetadata::LEFT] == std::vector<int> { 6, 6 });
+        REQUIRE(metadata.neighbourRanks[ModelMetadata::RIGHT].size() == 0);
+        REQUIRE(metadata.neighbourRanks[ModelMetadata::BOTTOM].size() == 0);
+        REQUIRE(metadata.neighbourRanks[ModelMetadata::TOP].size() == 0);
+    } else {
+        std::cerr << "only valid for 3 ranks" << std::endl;
+        exit(1);
+    }
```
Comment (joewallwork) [inline core/test/ModelMetadata_test.cpp]: Could this common part of the tests be pulled out into a function that's called in both places? It'd avoid duplication and reduce the likelihood of human errors when copy/pasting and reviewing.

File: core/test/ModelMetadata_test.cpp [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    // this metadata is specific to the periodic boundary conditions
+    if (test_rank == 0) {
+        // clang-format off
```
Comment (joewallwork) [inline core/test/ModelMetadata_test.cpp]: I'm guessing this is because of line length? Might be sufficient to put `using namespace std;` at the top.

File: core/test/ModelMetadata_test.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+#include <doctest/extensions/doctest_mpi.h>
+#include <iostream>
+
+#include "ModelMetadata.hpp"
+
+const std::string testFilesDir = TEST_FILES_DIR;
+const std::string filename = testFilesDir + "/paraGrid_test.nc";
+const std::string diagFile = "paraGrid_diag.nc";
+const std::string dateString = "2000-01-01T00:00:00Z";
```
Comment (joewallwork) [inline core/test/ModelMetadata_test.cpp]: Unused?

**Review by joewallwork — CHANGES_REQUESTED** [review]
Looks great, thanks @TomMelt. I drew a diagram and convinced myself that the numbers in the CDL files are consistent with the expected values in the tests. I have a couple of suggestions, the main one being a bit of refactoring to avoid duplication.

File: core/src/ModelMetadata.cpp [hunk 4]
```diff
@@ -39,6 +42,75 @@ void ModelMetadata::setMpiMetadata(MPI_Comm comm)
     MPI_Comm_rank(mpiComm, &mpiMyRank);
 }
 
+void ModelMetadata::readNeighbourData(netCDF::NcFile& ncFile)
+{
+    netCDF::NcGroup neighbourGroup(ncFile.getGroup(neighbourName));
+    std::string varName {};
+    for (auto edge : edges) {
+        size_t nStart {}; // start point in metadata arrays
+        size_t count {}; // number of elements to read from metadata arrays
```
Comment (timspainNERSC) [inline core/src/ModelMetadata.cpp]: Initialize numeric types using `= 0`, if indeed that is necessary (`nStart` and `count` seem to be set to other values before being read from).

File: core/test/ModelMetadata_test.cpp (hunk 2 above)
Comment (timspainNERSC) [reply core/test/ModelMetadata_test.cpp]: Or define a `typedef` (or `using`)

**Review by timspainNERSC — CHANGES_REQUESTED** [review]
A couple of minor edits.

**Comment by timspainNERSC** [comment]
Having looked over the changes to ModelMetadata.cpp again, I'm not sure this code should be in there. Perhaps even the grid metadata should be moved out. You don't have to stuff the MPI metadata into existing classes.

Perhaps we should meet to discuss the architecture of the MPI data now that we have more of a handle on what it is going to look like. Possibly including moving the existing coordinate data and functions that are currently in ModelMetadata.

**Comment by TomMelt** [comment]
> Having looked over the changes to ModelMetadata.cpp again, I'm not sure this code should be in there. Perhaps even the grid metadata should be moved out. You don't have to stuff the MPI metadata into existing classes.
> 
> Perhaps we should meet to discuss the architecture of the MPI data now that we have more of a handle on what it is going to look like. Possibly including moving the existing coordinate data and functions that are currently in ModelMetadata.

@timspainNERSC , yeah that makes sense. I will continue with my halo exchange tests for now and we can sort this before it merges into `develop`

**Comment by TomMelt** [comment]
This PR has been superseded by #744
